File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/DomainViewModel.java
package com.platdmit.simplecloudmanager.app.vm;

import com.platdmit.simplecloudmanager.domain.models.Domain;
import com.platdmit.simplecloudmanager.domain.repo.DomainBaseRepo;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;
import io.reactivex.rxjava3.processors.BehaviorProcessor;

public class DomainViewModel extends BaseViewModel {
    private static final String TAG = DomainViewModel.class.getSimpleName();
    private DomainBaseRepo mDomainRepo;
    private LiveData<Domain> mDomain;

    private final BehaviorProcessor<Domain> mContentProvider = BehaviorProcessor.create();

    public DomainViewModel(DomainBaseRepo domainBaseRepo, long id) {
        mDomainRepo = domainBaseRepo;

        //Fast fix for prevent overSubscription after resize
        mCompositeDisposable.add(mDomainRepo.getDomain(id).subscribe(mContentProvider::onNext));

        mDomain = LiveDataReactiveStreams.fromPublisher(mContentProvider);
    }

    public LiveData<Domain> getDomainLiveData() {
        return mDomain;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        mContentProvider.onComplete();
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/vm/DomainViewModel.kt
package com.platdmit.simplecloudmanager.vm

import androidx.hilt.Assisted
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import androidx.lifecycle.SavedStateHandle
import com.platdmit.domain.models.Domain
import com.platdmit.domain.repo.DomainBaseRepo
import com.platdmit.simplecloudmanager.states.DomainState
import io.reactivex.rxjava3.processors.BehaviorProcessor

class DomainViewModel
@ViewModelInject
constructor(
        private val domainBaseRepo: DomainBaseRepo,
        @Assisted private val savedStateHandle: SavedStateHandle
) : BaseViewModel<DomainState>() {
    val domainStateLiveData = LiveDataReactiveStreams.fromPublisher(stateProvider)
    val messageLiveData = LiveDataReactiveStreams.fromPublisher(messageProvider)

    init {
        stateProvider.onNext(DomainState.Loading)
    }

    fun setStateIntent(stateIntent: StateIntent){
        when(stateIntent){
            is StateIntent.SetDomainId -> {
                setActiveId(stateIntent.id)
            }
            is StateIntent.RefreshResult -> {}
        }
    }

    private fun setActiveId(id: Long){
        //Fast fix for prevent overSubscription after resize
        compositeDisposable.add(
                domainBaseRepo.getDomain(id).subscribe ({
                    stateProvider.onNext(DomainState.Success(it))
                },{
                    stateProvider.onNext(DomainState.Error)
                })
        )
    }

    sealed class StateIntent {
        data class SetDomainId(val id: Long) : StateIntent()
        object RefreshResult : StateIntent()
    }
}